### PR TITLE
fix: mutex deadlocks in PeerClient

### PR DESCRIPTION
**Pull request by miparnisari**
## Description

Close https://github.com/mailgun/gubernator/issues/222. 

The mutex deadlock is proven by the flaky `TestPeerClientShutdown`:

```
[25/02/24 7:09:39] ~/GitHub/gubernator (master) $ go test ./... -run=TestPeerClientShutdown -count=1 -race -timeout=10s
?   	github.com/mailgun/gubernator/v2/cmd/gubernator	[no test files]
?   	github.com/mailgun/gubernator/v2/cmd/gubernator-cli	[no test files]
?   	github.com/mailgun/gubernator/v2/cmd/gubernator-cluster	[no test files]
?   	github.com/mailgun/gubernator/v2/cmd/healthcheck	[no test files]
time="2024-02-25T19:09:44-03:00" level=info msg="Starting 12 Gubernator workers..." instance="127.0.0.1:9990"
time="2024-02-25T19:09:44-03:00" level=info msg="GRPC Listening on 127.0.0.1:9990 ..." instance="127.0.0.1:9990"
time="2024-02-25T19:09:44-03:00" level=info msg="HTTP Gateway Listening on 127.0.0.1:9980 ..." instance="127.0.0.1:9990"
time="2024-02-25T19:09:44-03:00" level=info msg="Starting 12 Gubernator workers..." instance="127.0.0.1:9991"
time="2024-02-25T19:09:44-03:00" level=info msg="GRPC Listening on 127.0.0.1:9991 ..." instance="127.0.0.1:9991"
time="2024-02-25T19:09:44-03:00" level=info msg="HTTP Gateway Listening on 127.0.0.1:9981 ..." instance="127.0.0.1:9991"
time="2024-02-25T19:09:44-03:00" level=info msg="Starting 12 Gubernator workers..." instance="127.0.0.1:9992"
time="2024-02-25T19:09:44-03:00" level=info msg="GRPC Listening on 127.0.0.1:9992 ..." instance="127.0.0.1:9992"
time="2024-02-25T19:09:44-03:00" level=info msg="HTTP Gateway Listening on 127.0.0.1:9982 ..." instance="127.0.0.1:9992"
time="2024-02-25T19:09:44-03:00" level=info msg="Starting 12 Gubernator workers..." instance="127.0.0.1:9993"
time="2024-02-25T19:09:44-03:00" level=info msg="GRPC Listening on 127.0.0.1:9993 ..." instance="127.0.0.1:9993"
time="2024-02-25T19:09:44-03:00" level=info msg="HTTP Gateway Listening on 127.0.0.1:9983 ..." instance="127.0.0.1:9993"
time="2024-02-25T19:09:44-03:00" level=info msg="Starting 12 Gubernator workers..." instance="127.0.0.1:9994"
time="2024-02-25T19:09:44-03:00" level=info msg="GRPC Listening on 127.0.0.1:9994 ..." instance="127.0.0.1:9994"
time="2024-02-25T19:09:44-03:00" level=info msg="HTTP Gateway Listening on 127.0.0.1:9984 ..." instance="127.0.0.1:9994"
time="2024-02-25T19:09:44-03:00" level=info msg="Starting 12 Gubernator workers..." instance="127.0.0.1:9995"
time="2024-02-25T19:09:44-03:00" level=info msg="GRPC Listening on 127.0.0.1:9995 ..." instance="127.0.0.1:9995"
time="2024-02-25T19:09:44-03:00" level=info msg="HTTP Gateway Listening on 127.0.0.1:9985 ..." instance="127.0.0.1:9995"
time="2024-02-25T19:09:44-03:00" level=info msg="Starting 12 Gubernator workers..." instance="127.0.0.1:9890"
time="2024-02-25T19:09:44-03:00" level=info msg="GRPC Listening on 127.0.0.1:9890 ..." instance="127.0.0.1:9890"
time="2024-02-25T19:09:44-03:00" level=info msg="HTTP Gateway Listening on 127.0.0.1:9880 ..." instance="127.0.0.1:9890"
time="2024-02-25T19:09:44-03:00" level=info msg="Starting 12 Gubernator workers..." instance="127.0.0.1:9891"
time="2024-02-25T19:09:44-03:00" level=info msg="GRPC Listening on 127.0.0.1:9891 ..." instance="127.0.0.1:9891"
time="2024-02-25T19:09:44-03:00" level=info msg="HTTP Gateway Listening on 127.0.0.1:9881 ..." instance="127.0.0.1:9891"
time="2024-02-25T19:09:44-03:00" level=info msg="Starting 12 Gubernator workers..." instance="127.0.0.1:9892"
time="2024-02-25T19:09:44-03:00" level=info msg="GRPC Listening on 127.0.0.1:9892 ..." instance="127.0.0.1:9892"
time="2024-02-25T19:09:44-03:00" level=info msg="HTTP Gateway Listening on 127.0.0.1:9882 ..." instance="127.0.0.1:9892"
time="2024-02-25T19:09:44-03:00" level=info msg="Starting 12 Gubernator workers..." instance="127.0.0.1:9893"
time="2024-02-25T19:09:44-03:00" level=info msg="GRPC Listening on 127.0.0.1:9893 ..." instance="127.0.0.1:9893"
time="2024-02-25T19:09:44-03:00" level=info msg="HTTP Gateway Listening on 127.0.0.1:9883 ..." instance="127.0.0.1:9893"
panic: test timed out after 10s
running tests:
	TestPeerClientShutdown (10s)
	TestPeerClientShutdown/Global (10s)


goroutine 566 [sync.RWMutex.Lock]:
sync.runtime_SemacquireRWMutex(0xc000fefedc?, 0xbc?, 0xc000fefec8?)
	/opt/homebrew/Cellar/go/1.21.6/libexec/src/runtime/sema.go:87 +0x28
sync.(*RWMutex).Lock(0xc000fefec8)
	/opt/homebrew/Cellar/go/1.21.6/libexec/src/sync/rwmutex.go:152 +0x88
github.com/mailgun/gubernator/v2.(*PeerClient).connect(0xc000fefe00, {0xc00150a1a0?, 0x1?})
	/Users/maria.inesparnisari/GitHub/gubernator/peer_client.go:123 +0x3b4
github.com/mailgun/gubernator/v2.(*PeerClient).getPeerRateLimitsBatch(0xc000fefe00, {0x1054e7e50, 0x1063ddd60}, 0xc00086a200)
	/Users/maria.inesparnisari/GitHub/gubernator/peer_client.go:299 +0x170
github.com/mailgun/gubernator/v2.(*PeerClient).GetPeerRateLimit(0xc000fefe00, {0x1054e7e50, 0x1063ddd60}, 0xc00086a200)
	/Users/maria.inesparnisari/GitHub/gubernator/peer_client.go:199 +0x4fc
github.com/mailgun/gubernator/v2_test.TestPeerClientShutdown.func1.1()
	/Users/maria.inesparnisari/GitHub/gubernator/peer_client_test.go:71 +0x12c
created by github.com/mailgun/gubernator/v2_test.TestPeerClientShutdown.func1 in goroutine 565
	/Users/maria.inesparnisari/GitHub/gubernator/peer_client_test.go:68 +0x3cc
```

Sometimes this test also gives "panic: send on close channel". It happens because when you call `shutdown`, we are closing the channel, but we are not preventing any further requests, which will then try to write to that channel.

## Testing

With this PR:

```shell
25/02/24 7:11:21] ~/GitHub/gubernator (fix-peer-client) $ go test ./... -run=TestPeerClientShutdown -count=10 -race -timeout=5s
?   	github.com/mailgun/gubernator/v2/cmd/gubernator	[no test files]
?   	github.com/mailgun/gubernator/v2/cmd/gubernator-cli	[no test files]
?   	github.com/mailgun/gubernator/v2/cmd/gubernator-cluster	[no test files]
?   	github.com/mailgun/gubernator/v2/cmd/healthcheck	[no test files]
ok  	github.com/mailgun/gubernator/v2	7.453s
ok  	github.com/mailgun/gubernator/v2/cluster	2.966s [no tests to run]
[25/02/24 7:11:37] ~/GitHub/gubernator (fix-peer-client) $
```

### CODE REVIEW

File: peer_client.go
Comment (miparnisari) [inline peer_client.go:147]: I didn't understand why every API attempts to create a new connection. Why not create one in `NewPeerClient`?

Comment (thrawn01) [reply peer_client.go:148]: We were new to GRPC and were used to having to manage the connection reconnect.  It's very strange that when using GRPC the user has to pass in the a new connection to make a GRPC client, but then the GRPC client is expected to restart the connection. Because we passed in the connection, we assumed (incorrectly) that we also had to manage the connect reconnect. Also, if I recall correctly, we didn't want to leave open connections to each peer for even if they were not used. But now we know, that is how GRPC is supposed to work.

I removed all of this crap when I removed GRPC in #206. For those still on V2, I never saw a need to refactor all this, if you plan on using v2 for a while and have an interest, you can refactor this if you wish!

**Comment by thrawn01** [comment]
This code is overly complex and I loath changing it. 😭 

I'm probably missing something obvious, but where are we demonstrating there is a a deadlock and where is the deadlock?

File: peer_client.go
Comment (thrawn01) [inline peer_client.go:468]: Shutdown/Close methods should always accept and `context.Cancel` as there may be some unknown or unforeseen reason the close cannot complete and it then hangs forever. This has happened in production on other services and it's always good to put a limit on how long you will wait for a shutdown to complete. This protects the service from having to be killed by the orchestration system without successfully saving it's internal state.  Shutting down a connection is a low priority thing, if for some reason the workgroup `Wait()`  hangs forever, then we can cancel the shutdown wait, and not interfere with normal shutdown of any service that uses gubernator as a library.

File: peer_client.go
Comment (miparnisari) [reply peer_client.go:148]: When you say "for those still on v2", what do you mean? Isn't everyone on v2?

In this pr I wanted to fix the flaky test but I realized that unless I make major changes (ie remove all the mutex logic), it's pretty much impossible. So I'm not sure how to proceed. Are there plans to release v3 soon? I haven't seen your PR in detail

**Comment by miparnisari** [comment]
> This code is overly complex and I loath changing it. 😭
> 
> I'm probably missing something obvious, but where are we demonstrating there is a a deadlock and where is the deadlock?

The deadlock is proven by the flaky test. Sometimes I get "panic: send on close channel". It happens because when you call shutdown, we are closing the channel, but we are not preventing any further requests, which will then try to write to that channy

File: peer_client.go
Comment (miparnisari) [reply peer_client.go:468]: I've never seen a shutdown function accept a context. E.g. sql.Close, or connection.Close. 

The context object is meant to be used when there is the potential chance that the request is cancelled by the user. This is not the case, Shutdown is invoked internally, not by a client. We should not use the context as a crutch to fix deadlock issues.

I can refactor this to avoid mutex altogether, but I would like to know if the work would be useful (given the v3 release)

Comment (Baliedge) [reply peer_client.go:468]: And to add, Mailgun plans to abandon this repo in favor of @thrawn01's new repo.  I suppose you can continue using the `/v2` flavor with the new repo until ready to migrate to `/v3`.

Either way, it looks like `/v2` will be deprecated very soon and gRPC support along with it.  I suggest not investing too much time into it.

Comment (miparnisari) [reply peer_client.go:468]: Thanks for the update! I can lend a hand in the new repo 😄

Comment (Baliedge) [reply peer_client.go:468]: > Out of interest, what was the biggest factor for the decision to drop gRPC?

@thrawn01 and I had some big conversations about this.  Reduced complexity and improved performance.  HTTP can perform more efficiently at scale than gRPC if done right.

This would be a good time for @thrawn01 to mention his work on Duh-RPC.

Comment (miparnisari) [reply peer_client.go:468]: > HTTP can perform more efficiently at scale than gRPC if done right.

What's the source for this? My understanding was this this in fact the opposite 😄 https://aws.amazon.com/compare/the-difference-between-grpc-and-rest/

> A gRPC API is better for these use cases:
> High-performance systems
> High data loads
> Real-time or streaming applications

Comment (miparnisari) [reply peer_client.go:468]: Another source: https://github.com/grpc-ecosystem/grpc-gateway#background

> gRPC is great -- it generates API clients and server stubs in many programming languages, it is fast, easy-to-use, **bandwidth-efficient** and its design is combat-proven by Google. However, you might still want to provide a traditional RESTful JSON API as well. Reasons can range from maintaining backward-compatibility, supporting languages or clients that are not well supported by gRPC, to simply maintaining the aesthetics and tooling involved with a RESTful JSON architecture.

**Comment by Baliedge** [comment]
@miparnisari Is this PR ready for merge?  Not concerned about the go-bench error.  But the PR is still in draft mode.

File: peer_client.go
Comment (thrawn01) [reply peer_client.go:468]: I intend to publish some more formal information about our GRPC journey at Mailgun. But as a quick overview, we are doing some very high performance and large scale stuff with GRPC and it hasn't lived up to all the hype.

In fact we have discovered that GRPC is slightly slower than using plan old HTTP/1 !!!!!!

As a result, I started working on a project called DUH-RPC, which is still in the works, but you can see the benchmarks comparing HTTP/1 with GRPC here https://github.com/duh-rpc/duh-go-benchmarks

The spec I'm working on is in progress but is here, feed back is welcome. 
https://github.com/duh-rpc/duh-go

I'll be traveling for the next 2 weeks, and I'm not always near my laptop, so I don't expect much movement on these projects until after. I still get notifications for these conversations, so I'll try to reply as I can.

File: peer_client.go
Comment (thrawn01) [reply peer_client.go:148]: "soon" is at least 3 weeks away as I'm traveling for the next 2 weeks. I leave it up to you if you wish to refactor the peer client to remove the mutexes and rely on GRPC for reconnect.

File: peer_client_test.go
Comment (miparnisari) [inline peer_client_test.go:86]: this is asserting that `true == true` 😅

File: peer_client_test.go
Comment (miparnisari) [inline peer_client_test.go:80]: this is not unwrapping errors, so it will never enter the 1st case

File: peer_client.go
Comment (miparnisari) [inline peer_client.go:80]: it could be blocking if we added `grpc.WithBlock()`, but it's not recommended (see https://pkg.go.dev/google.golang.org/grpc#WithBlock)

Comment (miparnisari) [reply peer_client.go]: reverted to be non-blocking to preserve existing behavior.

**Comment by Baliedge** [comment]
> @Baliedge i don't know what to do about the newly failing `TestLeakyBucketDivBug`. It seems to be failing because of the immediately preceding test `TestHealthCheck`.

Since `TestHealthCheck` causes a server side shutdown and restart, I wonder if it's not ready in time for the next test.

File: functional_test.go
Comment (Baliedge) [inline functional_test.go:1621]: Would prefer if we could poll with `testutil.UntilPass()` for healthy.  A fixed delay may still be flaky on slow GHA servers.

**Review by Baliedge — CHANGES_REQUESTED** [review]

File: functional_test.go
Comment (miparnisari) [reply functional_test.go]: updated

File: functional_test.go
Comment (Baliedge) [inline functional_test.go:44]: No need to expose `allPeers`.  You can use `cluster.GetPeers()`.

**Review by Baliedge — CHANGES_REQUESTED** [review]

File: peer_client.go
Comment (miparnisari) [inline peer_client.go:419]: not sure if we still want this 🤔

File: functional_test.go
Comment (miparnisari) [reply functional_test.go]: updated

File: peer_client.go
Comment (Baliedge) [reply peer_client.go:419]: We can punt on this.  It clears the client's error list and there could be a good reason to do this.

**Review by Baliedge — APPROVED** [review]
Nice

**Comment by miparnisari** [comment]
@Baliedge `make test` still fails locally 😭

**Comment by Baliedge** [comment]
@miparnisari I noticed after merging.  It does work on retry.  If you have an idea how to make it better, let's try it.

**Pull request merged**
